### DISCUSSION TITLE
ESMPy fails when creating a Grid for a high-resolution dataset

### DISCUSSION THREAD

**Discussion by JSAnandEOS** [esmf-org/esmf#263] [ESMPy] [answered]
### Requirements

- [X] Reviewed [ESMPy Documentation](https://earthsystemmodeling.org/esmpy/)
- [X] Searched [GitHub Discussions](https://github.com/orgs/esmf-org/discussions?discussions_q=)

### Affiliation(s)

University of Leicester

### ESMF Version

v8.6.1 (ESMF and ESMPy)

### Issue

I originally came across this problem when using xESMF (see [here](https://github.com/pangeo-data/xESMF/issues/380)), but thought I might get a better reply from this group.

I'm trying to regrid some rectilinear data from Sentinel-3 (~300 m resolution, global coverage) onto the EURO-CORDEX curvilinear grid (~12.5 km resolution, European coverage) using conservative regridding. However, even after clipping my Sentinel data to the 12 km grid extent, the size of the dataset is still very large (approx. 17000 lat x 39000 lon cells), which causes ESMPy to fail when creating a Grid:

```
import esmpy
esmpy.Grid(np.array([20000, 15000]),  staggerloc=esmpy.StaggerLoc.CENTER, coord_sys=esmpy.CoordSys.SPH_DEG)
```

```
File ~/miniconda3/envs/esmpy_test/lib/python3.12/site-packages/esmpy/util/esmpyarray.py:38, in ndarray_from_esmf(data, dtype, shape)
     33 else:
     34     buffer = np.core.multiarray.int_asbuffer(
     35         ct.addressof(data.contents), size)
---> 38 esmfarray = np.ndarray(tuple(shape[:]), constants._ESMF2PythonType[dtype],
     39                        buffer, order="F")
     41 return esmfarray

TypeError: buffer is too small for requested array
```

Is there a way around this in ESMPy, or would I need to use ESMF explicitly? I'm not sure how to write a netCDF file for a curvilinear grid including cell boundaries that ESMF would recognise (I couldn't find any examples online that worked with `ESMF_Regrid`).

### Autotag

@billsacks

**Comment by billsacks** [comment, marked as answer]
Thanks for reporting this. I was able to reproduce it. The problem arises because the memory used to represent your grid slightly exceeds the 32-bit integer limit (20000 * 15000 * 8 bytes = 2.4 billion, exceeding the 32-bit limit of ~ 2.1 billion). If you were able to clip your grid just a bit more, you could avoid this issue.

If you're comfortable installing ESMPy from source (https://earthsystemmodeling.org/esmpy_doc/release/latest/html/install.html#installing-esmpy-from-source), you can patch the code with this fix that I have just submitted:

https://github.com/esmf-org/esmf/pull/267

We will try to get this fix into the upcoming ESMF 8.7 release.

**Reply by billsacks** [reply]
> If you were able to clip your grid just a bit more, you could avoid this issue.

Actually, re-reading your original post, it sounds like your actual grid is larger than the one in the reproducer, so I'm guessing this suggestion of clipping it slightly more won't work for you.
